### ISSUE TITLE
"PayPal Express Checkout Token does not exist" but order created

### ISSUE THREAD

**Issue by bharanikonda** [magento/magento2#28916]
### Preconditions (*)

1. Magento 2.3.1
2. Enable PayPal Express checkout

### Steps to reproduce (*)

1. Try placing an order with PayPal express checkout
2. Sometimes it throws "PayPal Express Checkout Token does not exist" error, the order is getting created.

### Expected result (*)

1. If there is any token issue, then the order should not be created.
2.

### Actual result (*)

1. Customer is redirected to the shopping cart page, PayPal Express Checkout Token does not exist error is displayed but the order is created and a confirmation email sent to the customer
2. https://www.screencast.com/t/UR2et14XGe0w

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by swathishs** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @swathishs. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by engcom-Oscar** [comment]
Hello @bharanikonda 

Thank you for your report.
Unfortunately, I was not able to reproduce this issue on the latest `2.4-develop` code. Probably it was already fixed and will appear in the next release. Or we missed some steps or preconditions.

Could you please recheck it on clean Magento 2.4 installation and confirm?

**Comment by engcom-Alfa** [comment]
Hi @bharanikonda . Thank you for your report.
Unfortunately, we are not able to reproduce the issue following by scenarios from this issue and #29013.
Probably it was already fixed.
Could you describe in more detail the scenario in which the error `PayPal Express Checkout Token does not exist` occurs?
Thanks!

**Comment by rohit-sobhana** [comment]
Experiencing same issue in Magento 2.3.5-p1 .PayPal Express Order gets placed but with no billing Address. Happens occasionally. Following is the stack trace.
```
 report.CRITICAL: Exception message: PayPal Express Checkout Token does not exist.
Trace: <pre>#1 Magento\Paypal\Controller\Express\AbstractExpress\PlaceOrder->execute() called at [vendor/magento/framework/App/Action/Action.php:108]
#2 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#3 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#4 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#5 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callPlugins() called at [generated/code/Magento/Paypal/Controller/Express/PlaceOrder/Interceptor.php:26]
#6 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:162]
#7 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:99]
#8 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#9 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#10 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php:99]
#11 Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#12 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/fastly/magento2/Model/FrontControllerPlugin.php:129]
#13 Fastly\Cdn\Model\FrontControllerPlugin->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#14 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php:69]
#15 Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#16 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#17 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:26]
#18 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#19 Magento\Framework\App\Http->launch() called at [vendor/magento/framework/App/Bootstrap.php:261]
#20 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:40]
```

**Comment by HenKun** [comment]
Occassionally we face this same issue on M2.4.1, since we switched to use Redis. 

```
[2020-12-04 09:36:18] main.CRITICAL: Exception message: PayPal Express Checkout Token existiert nicht.
Trace: <pre>#1 Magento\Paypal\Controller\Express\AbstractExpress\PlaceOrder->execute() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#2 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callParent('execute', array()) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#3 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php:51]
#4 Magento\Framework\App\Action\Plugin\ActionFlagNoDispatchPlugin->aroundExecute(&Magento\Paypal\Controller\Express\PlaceOrder\Interceptor#0000000057f2e0de00000000358728d8#, &Closure#0000000057f2e0c900000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:135]
#5 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#6 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callPlugins('execute', array(), array(array('invalidate_expir...', 'customer-app-act...'))) called at [generated/code/Magento/Paypal/Controller/Express/PlaceOrder/Interceptor.php:23]
#7 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:111]
#8 Magento\Framework\App\Action\Action->dispatch(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#9 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callParent('dispatch', array(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#)) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#10 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#11 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->___callPlugins('dispatch', array(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#), array(array('Mirasvit_Seo::co...', 'ajaxcart-action-...', 'mstSeoContentPut...', 'mstSeoContentApp...', 'mstSeoContentApp...', 'tk1SeoContentApp...', 'mstSeoAutolinkAd...', 'mstSeoContentApp...'), array('catalog_app_acti...', 'contextPlugin'))) called at [generated/code/Magento/Paypal/Controller/Express/PlaceOrder/Interceptor.php:32]
#12 Magento\Paypal\Controller\Express\PlaceOrder\Interceptor->dispatch(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/App/FrontController.php:186]
#13 Magento\Framework\App\FrontController->processRequest(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#, &Magento\Paypal\Controller\Express\PlaceOrder\Interceptor#0000000057f2e0de00000000358728d8#) called at [vendor/magento/framework/App/FrontController.php:118]
#14 Magento\Framework\App\FrontController->dispatch(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#15 Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', array(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#)) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#16 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php:99]
#17 Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(&Magento\Framework\App\FrontController\Interceptor#0000000057f2e37500000000358728d8#, &Closure#0000000057f2e2d600000000358728d8#, &Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:135]
#18 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/litespeed/module-litemage/Model/App/FrontController/LitemagePlugin.php:54]
#19 Litespeed\Litemage\Model\App\FrontController\LitemagePlugin->aroundDispatch(&Magento\Framework\App\FrontController\Interceptor#0000000057f2e37500000000358728d8#, &Closure#0000000057f2e2d600000000358728d8#, &Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:135]
#20 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php:71]
#21 Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(&Magento\Framework\App\FrontController\Interceptor#0000000057f2e37500000000358728d8#, &Closure#0000000057f2e2d600000000358728d8#, &Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:135]
#22 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#23 Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', array(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#), array(array('mstCoreDispatch'))) called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:23]
#24 Magento\Framework\App\FrontController\Interceptor->dispatch(&Magento\Framework\App\Request\Http#0000000057f2e3c000000000358728d8#) called at [vendor/magento/framework/App/Http.php:116]
#25 Magento\Framework\App\Http->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:23]
#26 Magento\Framework\App\Http\Interceptor->launch() called at [vendor/magento/framework/App/Bootstrap.php:263]
#27 Magento\Framework\App\Bootstrap->run(&Magento\Framework\App\Http\Interceptor#0000000057f2e30800000000358728d8#) called at [pub/index.php:40]
```

The order gets created and the payment was successful.

This can further result in issue https://github.com/magento/magento2/issues/13952 where duplicated orders are created. (This already has a patch, see https://github.com/magento/magento2/issues/13952#issuecomment-711439385)

**Comment by sergey-fedosimov** [comment]
Magento 2.3.5-p2 with MDVA-31006. The issue still here.

**Comment by usuyukiso** [comment]
Sometimes reproduced in M2.4.1

**Comment by rb-ec** [comment]
This still looks to be an issue in 2.3.6-p1. Duplicate orders created when customer uses PayPal Express.

**Comment by Penchal-magento** [comment]
Hi @bharaniko:Unfortunately, not able to reproduce  on Magento ver. dev-2.4-develop code. Probably it was already fixed.

**Comment by engcom-Alfa** [comment]
Hi @bharaniko
Unfortunately, The issue is not reproducible in Magento ver. dev-2.4-develop code.
Kindly upgrade to the latest version, and process with it. Still if you find in the latest version you can report the new bug on the same.

**Comment by TDM-Charles** [comment]
We're seeing this issue in 2.4.2

**Comment by zahidhussain20** [comment]
We have also same issue, but order is created

**Comment by jalung** [comment]
I'm seeing this message also, but it seems to appear only in the case of a "bot" requesting a specific link:

`GET /paypal/express/cancel/ HTTP/2.0`

It causes an Exception in the logs M2.4.3-p1
`main.CRITICAL: Exception message: PayPal Express Checkout Token does not exist.
Trace: <pre>#1 Magento\Paypal\Controller\Express\AbstractExpress\Cancel->execute() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#2 Magento\Paypal\Controller\Express\Cancel\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#3 Magento\Paypal\Controller\Express\Cancel\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#4 Magento\Paypal\Controller\Express\Cancel\Interceptor->___callPlugins() called at [generated/code/Magento/Paypal/Controller/Express/Cancel/Interceptor.php:23]
#5 Magento\Paypal\Controller\Express\Cancel\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:111]
#6 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#7 Magento\Paypal\Controller\Express\Cancel\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#8 Magento\Paypal\Controller\Express\Cancel\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#9 Magento\Paypal\Controller\Express\Cancel\Interceptor->___callPlugins() called at [generated/code/Magento/Paypal/Controller/Express/Cancel/Interceptor.php:32]
#10 Magento\Paypal\Controller\Express\Cancel\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:245]
#11 Magento\Framework\App\FrontController->getActionResponse() called at [vendor/magento/framework/App/FrontController.php:212]
#12 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:147]
#13 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#14 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#15 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php:99]
#16 Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#17 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php:71]
#18 Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#19 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#20 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:23]
#21 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#22 Magento\Framework\App\Http->launch() called at [vendor/magento/framework/App/Bootstrap.php:264]
#23 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:29]`

**Comment by tiagofssampaio** [comment]
Just leaving here a comment because I was in the same boat as you guys.
Was happening to me in Safari only.

After a lot of digging I found this issue #26377 and in the comment section @HirokazuNishi with his [answer](https://github.com/magento/magento2/issues/26377#issuecomment-585497122) saved my week.

After you install https://github.com/Veriteworks/CookieFix, make sure to go to General > Web > Default Cookie Settings and set "SameSite" to "none".

Hope I saved yours too! 😄
